core/src/main/java/org/apache/calcite/plan/volcano/RelSubset.java

```diff
@@ -344,15 +344,29 @@ void propagateCostImprovements0(VolcanoPlanner planner, RelMetadataQuery mq,
       return;
     }
     try {
-      final RelOptCost cost = planner.getCost(rel, mq);
-      if (cost.isLt(bestCost)) {
-        LOGGER.trace("Subset cost improved: subset [{}] cost was {} now {}", this, bestCost, cost);
+      RelOptCost cost = planner.getCost(rel, mq);
+      boolean updateBest = cost.isLt(bestCost);
+
+      // Best rel's cost is increased, we need to search for new best rel
+      if (rel == best && bestCost.isLt(cost)) {
+        updateBest = true;
```

Comment:
Why we need a decision `rel == best` ? We should always update the best rel and cost if we found any. The problem here is that when and how this fix is triggered, can you give more explain ?

Reply:
We only need to update best when its own cost is increased. If this is any other rel node and its cost is larger than current best, we simply ignore (same behavior as today). I think I have explained the case when this could happen. Please check the descriptions and JIRA. Let me know if you have further question. Thanks.

Reply:
But in this for loop, you have replaced the best `rel` right ? This replaced rel may still not be the best one, i didn't find the meaningness to replace it, if you want to fix the cost val, just update the cost.

Reply:
In what case do you think the replaced rel may still not be the best one?

Reply:
The loop is to guarantee the new best would be the one with cheapest cost in current sub set. It could be a different node if the cost of the original best has increased. As I mentioned in the JIRA, this approach doesn't guarantee an optimal result but does make sure the memo is consistent. To get optimal result, we either need to make sure RelSubset row count doesn't change (it shouldn't change in theory), or we have multiple best node candidates in a subset. But the change would be too drastic and the runtime overhead is non trivial.